Optimize TransformationMatrix inverse on ARM64

Ahmad-S792 opened the pull request:
#### 6b8fc20df8d4ed0fb93e5ad39bc0ea22599743d9
<pre>
Optimize TransformationMatrix inverse on ARM64
<a href="https://rdar.apple.com/147554888">rdar://147554888</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=290156">https://bugs.webkit.org/show_bug.cgi?id=290156</a>

Reviewed by Yusuke Suzuki.

This is another attempt of 260468@main, which was reverted in 260765@main.
In this attempt, it merges following two commits:

<a href="https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5">https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5</a> &amp;
<a href="https://chromium.googlesource.com/chromium/src/+/d5b820e21463e127cfac509d7bfdd27bb894b1ca">https://chromium.googlesource.com/chromium/src/+/d5b820e21463e127cfac509d7bfdd27bb894b1ca</a>

In previous attempt, it had typo error in `v24` and `v25` registers, which
would have led to unexpected values in calls to inverse function. Besides
following, it also don&apos;t use previous `rdet` in limited scope and have it
outside as `inverseDeterminant`.

From original testing by Chris, it is performance optimization of about
20% tested on M1 Pro machine with 100k matrices of random double values.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::adjoint):
(WebCore::inverse):
(WebCore::TransformationMatrix::multiply):

Canonical link: <a href="https://commits.webkit.org/292492@main">https://commits.webkit.org/292492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/088646b8ecc9292b9e401fc9a950373f4948c3ad

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96155 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15769 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5730 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101220 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46685 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98200 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16064 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24202 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73302 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30541 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99158 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12070 "Found 1 new test failure: fast/forms/ios/focus-input-via-button.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86854 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53639 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11810 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4615 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45999 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81919 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103245 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23222 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16913 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82339 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23473 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82875 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81715 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20522 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26330 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3758 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16583 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23185 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/28340 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22844 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26324 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24585 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
```diff
@@ -1168,7 +1283,7 @@ TransformationMatrix& TransformationMatrix::zoom(double zoomFactor)
 TransformationMatrix& TransformationMatrix::multiply(const TransformationMatrix& mat)
 {
-#if CPU(ARM64) && defined(_LP64)
+#if CPU(ARM64) && defined(ADDRESS64)
```
This needs to be `CPU(ADDRESS64)`.

Ahmad-S792 replied:
Fixed in [088646b](https://github.com/WebKit/WebKit/commit/088646b8ecc9292b9e401fc9a950373f4948c3ad).

Ahmad-S792 committed:
Optimize TransformationMatrix inverse on ARM64
rdar://147554888
https://bugs.webkit.org/show_bug.cgi?id=290156

Reviewed by Yusuke Suzuki.

This is another attempt of 260468@main, which was reverted in 260765@main.
In this attempt, it merges following two commits:

https://chromium.googlesource.com/chromium/blink/+/7b03a53c2c7f24fb1dd79c0e7f4b7fad520875a5 &
https://chromium.googlesource.com/chromium/src/+/d5b820e21463e127cfac509d7bfdd27bb894b1ca

In previous attempt, it had typo error in `v24` and `v25` registers, which
would have led to unexpected values in calls to inverse function. Besides
following, it also don't use previous `rdet` in limited scope and have it
outside as `inverseDeterminant`.

From original testing by Chris, it is performance optimization of about
20% tested on M1 Pro machine with 100k matrices of random double values.

* Source/WebCore/platform/graphics/transforms/TransformationMatrix.cpp:
(WebCore::adjoint):
(WebCore::inverse):
(WebCore::TransformationMatrix::multiply):

Canonical link: https://commits.webkit.org/292492@main
